### PR TITLE
Copy the DOM XPath interfaces from the WHATWG wiki

### PR DESCRIPTION
**Pull request by foolip**
https://wiki.whatwg.org/wiki/DOM_XPath

Related: https://github.com/whatwg/dom/issues/67

***
<a href="https://whatpr.org/dom/763.html" title="Last updated on Aug 30, 2019, 8:19 AM UTC (5f1b1e0)">Preview</a> | <a href="https://whatpr.org/dom/763/a93f9f8...5f1b1e0.html" title="Last updated on Aug 30, 2019, 8:19 AM UTC (5f1b1e0)">Diff</a>

### CODE REVIEW

**Comment by foolip** [comment]
Tests using this IDL already exist:
https://wpt.fyi/results/domxpath/interfaces.tentative.html?run_id=258200004&run_id=264060011&run_id=250500005&run_id=266120012

They show that all of this is already implemented, the failures in Edge and Safari are because of minor differences in `length` properties, what exception is thrown, and so on.

File: dom.bs [hunk 1]
```diff
@@ -9818,6 +9818,88 @@ the given value.
...
+<h3 id=mixin-xpathevaluatorbase>Mixin {{XPathEvaluatorBase}}</h3>
+
+<pre class=idl>
+callback interface XPathNSResolver {
+  DOMString? lookupNamespaceURI(DOMString? prefix);
+};
+
+interface mixin XPathEvaluatorBase {
+  [NewObject] XPathExpression createExpression(DOMString expression,
+                                               optional XPathNSResolver? resolver);
```
Comment (gsnedders) [inline dom.bs]:
```suggestion
                                               optional XPathNSResolver? resolver = null);
```

File: dom.bs [hunk 2]
```diff
@@ -9818,6 +9818,88 @@ the given value.
...
+};
+</pre>
+
+<h3 id=interface-xpathexpression>Interface {{XPathExpression}}</h3>
+
+<pre class=idl>
+[Exposed=Window]
+interface XPathExpression {
+  XPathResult evaluate(Node contextNode,
+                       optional unsigned short type,
```
Comment (gsnedders) [inline dom.bs]:
```suggestion
                       optional unsigned short type = 0 /* ANY_TYPE */,
```

File: dom.bs [hunk 3]
```diff
@@ -9818,6 +9818,88 @@ the given value.
...
+                       optional XPathResult? result);
```
Comment (gsnedders) [inline dom.bs]:
```suggestion
                       optional XPathResult? result = null);
```

File: dom.bs [hunk 4]
```diff
@@ -9818,6 +9818,88 @@ the given value.
...
+  DOMString? lookupNamespaceURI(DOMString? prefix);
+};
+
+interface mixin XPathEvaluatorBase {
+  [NewObject] XPathExpression createExpression(DOMString expression,
+                                               optional XPathNSResolver? resolver);
+  XPathNSResolver createNSResolver(Node nodeResolver);
+  XPathResult evaluate(DOMString expression,
+                       Node contextNode,
+                       optional XPathNSResolver? resolver,
```
Comment (gsnedders) [inline dom.bs]:
```suggestion
                       optional XPathNSResolver? resolver = null,
```

File: dom.bs [hunk 5]
```diff
@@ -9818,6 +9818,88 @@ the given value.
...
+                       optional unsigned short type,
```
Comment (gsnedders) [inline dom.bs]:
```suggestion
                       optional unsigned short type = 0 /* ANY_TYPE */,
```

File: dom.bs [hunk 6]
```diff
@@ -9818,6 +9818,88 @@ the given value.
...
+                       optional XPathResult? result);
```
Comment (gsnedders) [inline dom.bs]:
```suggestion
                       optional XPathResult? result = null);
```

File: dom.bs [hunk 7]
```diff
@@ -9818,6 +9818,88 @@ the given value.
 
 
 
+<h2 id=xpath>XPath</h2>
+
+<cite>DOM Level 3 XPath</cite> defined an API for evaluating <cite>XPath 1.0</cite> expressions.
+[[DOM-Level-3-XPath]] [[XPath]] These APIs are widely implemented, but
+<cite>DOM Level 3 XPath</cite> is no longer maintained. The DOM Standard does not attempt to
+define any of the behavior, but the interface definitions are maintained here so that they can
+be updated when Web IDL changes. [[WEBIDL]]
```
Comment (gsnedders) [inline dom.bs]: There's more changes than just that, some of them simplify the API by making arguments optional.

Let's also add something, possibly just normative, saying:

 * `XPathException` is gone (just use `DOMException`)
 * The query is matched against the DOM, and therefore contrary to the XPath 1.0 data model the root element has a parent (the Document) and text nodes can be adjacent to one another. [wilful violation]

**Comment by foolip** [comment]
I'll polish this up next week.

**Comment by foolip** [comment]
I didn't get to this this week and I'm OOO tomorrow. Snoozing it until next week but won't be offended if someone takes it over.

**Comment by foolip** [comment]
@annevk I pushed the changes. However, `class=issue` doesn't seem to be a thing, so I used `class=warning`. Don't love the way it looks, is there a better class?

**Comment by zcorpan** [comment]
@foolip `class=XXX`

**Comment by foolip** [comment]
@zcorpan thanks, trying that instead

**Comment by annevk** [comment]
This is blocked by https://github.com/w3ctag/promises-guide/issues/27 for now. I adjusted the wording a bit to make it clear that defining these is still a goal, albeit perhaps with less priority than some would like.

**Comment by foolip** [comment]
https://github.com/w3ctag/promises-guide/issues/27 is resolved, Travis now happy.

**Pull request merged**